posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,196 @@
+The common architecture of SciPy, scikit-learn, and scikit-image is that they use NumPy for array operations, and critical-to-performance parts are written using compiled code extensions. Unfortunately, compiled extensions rely on NumPy's internal memory representation and are restricted to CPU computations. For these reasons, Python code with compiled extensions is not compatible with NumPy's dispatching capabilities; it would not know how to deal with a GPU array. It's also very challenging to rewrite the libraries to use pure NumPy without sacrificing the performance. To solve the problem we should introduce a similar to NumPy dispatching mechanism that would be used for the modules that rely on compiled extensions.
+
+Let's imagine some SciPy module that equally depends on compiled code and NumPy functions. Removing all forced conversions to NumPy arrays inside the module would open the "Dispatcher" path, but only half of the SciPy module's functionality would work for generic arrays. In the diagram, the NumPy Dispatcher is an abstract mechanism, concretely it could be an `__array_function/ufunc__`-based one for older versions of NumPy or `__array_namespace__` - for NumPy 1.22+.
+
+<p align="center">
+    <img
+     alt="A schema using NumPy dispatcher."
+     src="/images/2021/11/scipy-numpy-dispatch.png">
+    <i>A SciPy module with NumPy "Dispatcher".</i>
+</p>
+
+<!-- ![numpy dispatch schema](https://drive.google.com/uc?export=view&id=1-PySRskT3-76r_KiiKGfkC485dT6uKoV) -->
+
+For SciPy and similar libraries, we need to follow NumPy's example and add our own dispatching layer for the API to allow users to develop generic code that works with arrays from different libraries. Then it would be possible to directly reroute to specific array implementations bypassing compiled code paths.
+
+<p align="center">
+    <img
+     alt="A schema using SciPy dispatcher."
+     src="/images/2021/11/scipy-numpy-dispatch.png">
+    <i>A SciPy module with its own "Dispatcher".</i>
+</p>
+
+<!-- ![scipy.ndimage dispatch schema](https://drive.google.com/uc?export=view&id=1-ZoVxenDufJuUwcONV_mFY0NFNY3wOCq) -->
+
+## Goals, wishes, and constraints
+
+Goal: separate the interface from the implementation and let the dispatching system help users and library authors write generic code that works with different kinds of arrays.
+
+The new dispatching system implemented for SciPy and scikits should:
+* make the API extendable and add the possibility to support alternative array libraries, not just NumPy,
+* provide a way to select backend for the same array type (example: `scipy.fft` + alternative CPU FFT libraries, scikit-learn + [Intel optimized implementation](https://github.com/intel/scikit-learn-intelex)),
+* be able to extend function signature in a backward-compatible way on the base library side (all alternative implementations should not break after the change),
+* have as low overhead as possible.
```

Comment:
```suggestion
* make the API extendable and add the possibility to support alternative array libraries, not just NumPy
* provide a way to select backend for the same array type (example: `scipy.fft` + alternative CPU FFT libraries, scikit-learn + [Intel optimized implementation](https://github.com/intel/scikit-learn-intelex))
* be able to extend function signature in a backward-compatible way on the base library side (all alternative implementations should not break after the change)
* have as low overhead as possible
```
Personal inclination to not use "," in these bullets. Not a big deal though. Feel free to ignore if you feel the other way.